File: app-web/vela-wallet/src/lib/settings/core/network-admin-executor.test.ts
/**
 * The web network_admin executor (spec 024 T018): one op ↔ one call, stored
 * bytes stay Expo-compatible, the failure twin answers every operation.
 * Transports are not exercised here — the probe bodies are ports and the
 * fetch they ride is mocked shut; what is pinned is shape translation.
 */
import { beforeEach, describe, expect, it, vi } from 'vitest';
import type { NetEffect } from './network-admin-types';

const kv = new Map<string, string>();
vi.mock('$lib/services/storage', () => ({
	getItem: vi.fn(async (key: string) => kv.get(key) ?? null),
	setItem: vi.fn(async (key: string, value: string) => void kv.set(key, value)),
	removeItem: vi.fn(async (key: string) => void kv.delete(key))
}));

const endpointsStore: { record: Record<string, string> } = { record: {} };
vi.mock('$lib/onboarding/core/storage', () => ({
	loadServiceEndpoints: vi.fn(() => endpointsStore.record),
	saveServiceEndpoints: vi.fn((record: Record<string, string>) => {
		endpointsStore.record = record;
	})
}));

const indexRows: { rows: unknown[] } = { rows: [] };
vi.mock('$lib/services/chain-registry', () => ({
	loadSearchIndex: vi.fn(async () => indexRows.rows),
	fetchRawChainData: vi.fn(async () => null)
}));

import {
	executeNetworkAdminOperation,
	networkAdminOperationFailure
} from './network-admin-executor';
import { saveServiceEndpoints } from '$lib/onboarding/core/storage';

const effect = (operation: NetEffect['operation']): NetEffect => ({ id: 1, operation });

const STORED_CUSTOM = {
	id: 'xlayer',
	displayName: 'X Layer',
	chainId: 196,
	iconLabel: 'X',
	iconColor: '#fff',
	iconBg: '#000',
	logoURL: '',
	isL2: true,
	rpcURL: 'https://rpc.xlayer.tech',
	explorerURL: 'https://www.oklink.com/xlayer',
	bundlerURL: '',
	nativeSymbol: 'OKB',
	addedAt: '2026-01-01T00:00:00.000Z'
};

beforeEach(() => {
	kv.clear();
	endpointsStore.record = {};
	vi.clearAllMocks();
});

describe('read_store', () => {
	it('translates the stored camelCase records to the wire', async () => {
		kv.set('vela.customNetworks', JSON.stringify([STORED_CUSTOM]));
		kv.set('vela.networkConfig', JSON.stringify([{ chainId: 1, rpcURL: 'https://rpc.example' }]));
		kv.set('vela.rpcProviders', JSON.stringify({ alchemy: 'key-a' }));
		endpointsStore.record = { passkeyIndexURL: 'https://idx.example' };

		const result = await executeNetworkAdminOperation(effect({ type: 'read_store' }));
		expect(result.type).toBe('store_loaded');
		if (result.type !== 'store_loaded') return;
		expect(result.custom_networks[0]).toMatchObject({
			id: 'xlayer',
			display_name: 'X Layer',
			chain_id: 196,
			rpc_url: 'https://rpc.xlayer.tech',
			native_symbol: 'OKB',
			added_at_iso: '2026-01-01T00:00:00.000Z'
		});
		// Missing stored fields coerce, never reject (a rejected StoreLoaded
		// would strand the core unloaded forever).
		expect(result.network_configs[0]).toEqual({
			chain_id: 1,
			rpc_url: 'https://rpc.example',
			explorer_url: '',
			bundler_url: ''
		});
		expect(result.provider_keys).toEqual({ alchemy: 'key-a', drpc: null, ankr: null });
		// Absent endpoint fields stay absent (null) — the core applies defaults.
		expect(result.endpoints).toEqual({
			ethereum_data_url: null,
			passkey_index_url: 'https://idx.example',
			bundler_service_url: null,
			fiat_rates_url: null
		});
	});

	it('reads junk as nothing-configured, still loaded', async () => {
		kv.set('vela.customNetworks', '{not json');
		const result = await executeNetworkAdminOperation(effect({ type: 'read_store' }));
		expect(result.type).toBe('store_loaded');
		if (result.type === 'store_loaded') expect(result.custom_networks).toEqual([]);
	});
});

describe('writes', () => {
	it('write_custom_networks stores the Expo camelCase shape, byte-compatible', async () => {
		const wire = {
			id: 'xlayer',
			display_name: 'X Layer',
			chain_id: 196,
			icon_label: 'X',
			icon_color: '#fff',
			icon_bg: '#000',
			logo_url: '',
			is_l2: true,
			rpc_url: 'https://rpc.xlayer.tech',
			explorer_url: 'https://www.oklink.com/xlayer',
			bundler_url: '',
			native_symbol: 'OKB',
			added_at_iso: '2026-01-01T00:00:00.000Z'
		};
		const result = await executeNetworkAdminOperation(
			effect({ type: 'write_custom_networks', networks: [wire] })
		);
		expect(result).toEqual({ type: 'written' });
		expect(JSON.parse(kv.get('vela.customNetworks')!)).toEqual([STORED_CUSTOM]);
	});

	it('write_service_endpoints goes through the onboarding module (D3a)', async () => {
		await executeNetworkAdminOperation(
			effect({
				type: 'write_service_endpoints',
				endpoints: {
					ethereum_data_url: 'https://data.example',
					passkey_index_url: 'https://idx.example',
					bundler_service_url: 'https://relay.example',
					fiat_rates_url: 'https://fx.example'
				}
			})
		);
		expect(saveServiceEndpoints).toHaveBeenCalledWith({
			ethereumDataURL: 'https://data.example',
			passkeyIndexURL: 'https://idx.example',
			bundlerServiceURL: 'https://relay.example',
			fiatRatesURL: 'https://fx.example'
		});
	});

	it('write_rpc_providers drops cleared keys instead of storing null', async () => {
		await executeNetworkAdminOperation(
			effect({ type: 'write_rpc_providers', keys: { alchemy: 'k', drpc: null, ankr: null } })
		);
		expect(JSON.parse(kv.get('vela.rpcProviders')!)).toEqual({ alchemy: 'k' });
	});
});

describe('platform-absent caches are acknowledged, never skipped', () => {
	it('invalidate_pools answers', async () => {
		expect(
			await executeNetworkAdminOperation(effect({ type: 'invalidate_pools', chain_id: null }))
		).toEqual({ type: 'invalidated' });
	});
	it('clear_bundler_cache answers', async () => {
		expect(
			await executeNetworkAdminOperation(effect({ type: 'clear_bundler_cache', chain_id: 1 }))
		).toEqual({ type: 'bundler_cache_cleared' });
	});
});

describe('the failure twin', () => {
	it('read_store failure reads as nothing-configured, still loaded', () => {
		const result = networkAdminOperationFailure(effect({ type: 'read_store' }));
		expect(result).toMatchObject({ type: 'store_loaded', custom_networks: [] });
	});
	it('probe_rpc failure answers the probed shape with an unknown chain id', () => {
		const result = networkAdminOperationFailure(
			effect({ type: 'probe_rpc', url: 'https://rpc.example' })
		);
		expect(result).toEqual({
			type: 'probed',
			url: 'https://rpc.example',
			reported_chain_id: null,
			latency_ms: 0
		});
	});
	it('fetch_service_health failure lands in the failed body', () => {
		const result = networkAdminOperationFailure(
			effect({ type: 'fetch_service_health', field: 'passkey_index', base_url: 'https://x' })
		);
		expect(result).toMatchObject({ type: 'service_health', body: { type: 'failed' } });
	});
});

describe('fetch_search_index', () => {
	it('drops a row whose chain id the core cannot represent instead of poisoning the whole index', async () => {
		indexRows.rows = [
			{
				chainId: 1,
				name: 'Ethereum',
				shortName: 'eth',
				nativeCurrencySymbol: 'ETH',
				hasLogo: true
			},
			// The live index's first over-u32 id: one row, and serde used to refuse ALL of them.
			{ chainId: 7078815900, name: 'Mezo Matsnet', shortName: 'mezo', nativeCurrencySymbol: 'BTC' },
			{ chainId: '100', name: 'Gnosis', shortName: 'gno', nativeCurrencySymbol: 'XDAI' },
			{ chainId: -5, name: 'Nope' },
			{ chainId: 1.5, name: 'Nope' },
			{ chainId: 'abc', name: 'Nope' }
		];
		const result = await executeNetworkAdminOperation(effect({ type: 'fetch_search_index' }));
		expect(result.type).toBe('search_index');
		if (result.type !== 'search_index') throw new Error('unreachable');
		expect(result.chains.map((c) => c.chain_id)).toEqual([1, 100]);
		expect(result.chains[0]).toEqual({
			chain_id: 1,
			name: 'Ethereum',
			short_name: 'eth',
			native_currency_symbol: 'ETH',
			has_logo: true
		});
	});
});

File: app-android/vela-wallet/app/src/test/java/app/getvela/wallet/NetworkAdminExecutorTest.kt
package app.getvela.wallet

import app.getvela.wallet.core.data.KeyValueStore
import app.getvela.wallet.feature.settings.core.NetCustomNetwork
import app.getvela.wallet.feature.settings.core.NetEndpointField
import app.getvela.wallet.feature.settings.core.NetHealthBody
import app.getvela.wallet.feature.settings.core.NetNetworkConfig
import app.getvela.wallet.feature.settings.core.NetOperation
import app.getvela.wallet.feature.settings.core.NetProviderKeys
import app.getvela.wallet.feature.settings.core.NetServiceEndpoints
import app.getvela.wallet.feature.settings.core.NetShellResult
import app.getvela.wallet.feature.settings.core.NetworkAdminExecutor
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The `network_admin` shell contract
 * (`specs/040-android-live-shell/contracts/shell-operations.md`).
 *
 * Two things this file is really about:
 *
 * 1. **The stored bytes are a cross-client contract.** The camelCase keys
 *    asserted below are what the web and desktop wallets read. A field written
 *    as `rpc_url` instead of `rpcURL` gives this device a network its siblings
 *    cannot see, and nothing anywhere reports a fault.
 * 2. **Nine operations are deliberately fail-closed** until spec 041, and each
 *    must answer the shape the core's failure twin defines — never a guess. An
 *    empty search index is not "no such chain"; a null code is not "not a
 *    contract".
 */
class NetworkAdminExecutorTest {

    private fun executor(store: KeyValueStore = FakeStore()) = NetworkAdminExecutor(store)

    // -- read_store ----------------------------------------------------------

    @Test
    fun anEmptyDeviceLoadsNothingConfigured() = runBlocking {
        val loaded = executor().perform(NetOperation.ReadStore) as NetShellResult.StoreLoaded
        assertTrue(loaded.custom_networks.isEmpty())
        assertTrue(loaded.network_configs.isEmpty())
        assertNull(loaded.endpoints.bundler_service_url)
        assertNull(loaded.provider_keys.alchemy)
    }

    @Test
    fun storedRecordsAreReadInTheOtherClientsShapes() = runBlocking {
        val store = FakeStore(
            mapOf(
                KeyValueStore.Keys.CUSTOM_NETWORKS to JSONArray().put(
                    JSONObject()
                        .put("id", "custom-7777")
                        .put("displayName", "Testnet")
                        .put("chainId", 7777)
                        .put("iconLabel", "T")
                        .put("iconColor", "#fff")
                        .put("iconBg", "#000")
                        .put("logoURL", "")
                        .put("isL2", true)
                        .put("rpcURL", "https://rpc.example")
                        .put("explorerURL", "https://scan.example")
                        .put("bundlerURL", "https://bundler.example")
                        .put("nativeSymbol", "TST")
                        .put("addedAt", "2026-09-05T00:00:00Z"),
                ).toString(),
                KeyValueStore.Keys.NETWORK_CONFIG to JSONArray().put(
                    JSONObject()
                        .put("chainId", 1)
                        .put("rpcURL", "https://eth.example")
                        .put("explorerURL", "")
                        .put("bundlerURL", ""),
                ).toString(),
                KeyValueStore.Keys.SERVICE_ENDPOINTS to
                    JSONObject().put("bundlerServiceURL", "https://b.example").toString(),
                KeyValueStore.Keys.RPC_PROVIDERS to
                    JSONObject().put("alchemy", "key-1").toString(),
            ),
        )

        val loaded = executor(store).perform(NetOperation.ReadStore) as NetShellResult.StoreLoaded

        assertEquals(1, loaded.custom_networks.size)
        assertEquals("Testnet", loaded.custom_networks[0].display_name)
        assertEquals(7777L, loaded.custom_networks[0].chain_id)
        assertTrue(loaded.custom_networks[0].is_l2)
        assertEquals("2026-09-05T00:00:00Z", loaded.custom_networks[0].added_at_iso)
        assertEquals(1L, loaded.network_configs[0].chain_id)
        assertEquals("https://b.example", loaded.endpoints.bundler_service_url)
        // Absent stays absent — the core applies the defaults, not the shell.
        assertNull(loaded.endpoints.fiat_rates_url)
        assertEquals("key-1", loaded.provider_keys.alchemy)
        assertNull(loaded.provider_keys.drpc)
    }

    @Test
    fun junkReadsAsEmptyRatherThanStrandingTheCore() = runBlocking {
        // The core would never leave `loaded` if `store_loaded` failed to
        // arrive, and every later write would be silently dropped. So garbage
        // coerces rather than throwing.
        val store = FakeStore(
            mapOf(
                KeyValueStore.Keys.CUSTOM_NETWORKS to "{not json at all",
                KeyValueStore.Keys.NETWORK_CONFIG to "\"a string, not an array\"",
                KeyValueStore.Keys.RPC_PROVIDERS to "[]",
            ),
        )
        val loaded = executor(store).perform(NetOperation.ReadStore) as NetShellResult.StoreLoaded
        assertTrue(loaded.custom_networks.isEmpty())
        assertTrue(loaded.network_configs.isEmpty())
        assertNull(loaded.provider_keys.ankr)
    }

    // -- the writes ----------------------------------------------------------

    @Test
    fun aCustomNetworkIsStoredInCamelCase() = runBlocking {
        val store = FakeStore()
        val network = NetCustomNetwork(
            id = "custom-42",
            display_name = "Forty Two",
            chain_id = 42,
            icon_label = "F",
            icon_color = "#111",
            icon_bg = "#222",
            logo_url = "https://logo.example",
            is_l2 = false,
            rpc_url = "https://rpc.example",
            explorer_url = "https://scan.example",
            bundler_url = "https://bundler.example",
            native_symbol = "FTT",
            added_at_iso = "2026-09-05T10:00:00Z",
        )

        val result = executor(store).perform(NetOperation.WriteCustomNetworks(listOf(network)))
        assertEquals(NetShellResult.Written, result)

        val stored = JSONArray(store.values[KeyValueStore.Keys.CUSTOM_NETWORKS]).getJSONObject(0)
        // The exact keys the sibling clients read.
        assertEquals("Forty Two", stored.getString("displayName"))
        assertEquals(42, stored.getInt("chainId"))
        assertEquals("https://rpc.example", stored.getString("rpcURL"))
        assertEquals("https://logo.example", stored.getString("logoURL"))
        assertEquals("2026-09-05T10:00:00Z", stored.getString("addedAt"))
        assertFalse(stored.getBoolean("isL2"))
    }

    @Test
    fun serviceEndpointsAreWrittenWhole() = runBlocking {
        val store = FakeStore()
        executor(store).perform(
            NetOperation.WriteServiceEndpoints(
                NetServiceEndpoints(
                    ethereum_data_url = "https://data.example",
                    passkey_index_url = "https://index.example",
                    bundler_service_url = "https://bundler.example",
                    fiat_rates_url = "https://rates.example",
                ),
            ),
        )
        val stored = JSONObject(store.values[KeyValueStore.Keys.SERVICE_ENDPOINTS])
        // The same key onboarding's endpoint sheet reads — one storage home.
        assertEquals("https://index.example", stored.getString("passkeyIndexURL"))
        assertEquals(4, stored.length())
    }

    @Test
    fun aClearedProviderKeyIsRemovedNotEmptied() = runBlocking {
        // The core's invariant ⑦. "" and absent are different: one falls back
        // to the public endpoint, the other fails every request with a key that
        // is not a key.
        val store = FakeStore()
        executor(store).perform(
            NetOperation.WriteRpcProviders(NetProviderKeys(alchemy = "abc", drpc = null)),
        )
        val stored = JSONObject(store.values[KeyValueStore.Keys.RPC_PROVIDERS])
        assertEquals("abc", stored.getString("alchemy"))
        assertFalse("a null key must not be stored at all", stored.has("drpc"))
    }

    @Test
    fun aRefusedWriteStillAnswers() = runBlocking {
        val store = FakeStore()
        store.refuseWrites = true
        assertEquals(
            NetShellResult.Written,
            executor(store).perform(NetOperation.WriteNetworkConfigs(listOf(NetNetworkConfig(1, "", "", "")))),
        )
        assertTrue(store.values.isEmpty())
    }

    // -- the fail-closed nine ------------------------------------------------

    @Test
    fun everyNetworkOperationAnswersItsUnknownShape() = runBlocking {
        val net = executor()

        assertEquals(
            NetShellResult.SearchIndex(emptyList()),
            net.perform(NetOperation.FetchSearchIndex),
        )
        assertEquals(
            NetShellResult.ChainInfo(8453, null),
            net.perform(NetOperation.FetchChainInfo(8453)),
        )
        assertEquals(
            NetShellResult.Probed("https://rpc", null, 0),
            net.perform(NetOperation.ProbeRpc("https://rpc")),
        )
        assertEquals(
            NetShellResult.Reachable("https://rpc", false, 0),
            net.perform(NetOperation.ProbeReachable("https://rpc")),
        )
        assertEquals(
            NetShellResult.Code("https://rpc", "0xabc", null),
            net.perform(NetOperation.RpcGetCode("https://rpc", "0xabc")),
        )
        assertEquals(
            NetShellResult.P256Call("https://rpc", null),
            net.perform(NetOperation.RpcCallP256("https://rpc")),
        )
        assertEquals(
            NetShellResult.ServiceHealth(NetEndpointField.FiatRates, NetHealthBody.Failed, 0),
            net.perform(
                NetOperation.FetchServiceHealth(NetEndpointField.FiatRates, "https://rates"),
            ),
        )
        assertEquals(
            NetShellResult.FiatRates(NetHealthBody.Failed, 0),
            net.perform(NetOperation.FetchFiatRates("https://rates")),
        )
        assertEquals(NetShellResult.Invalidated, net.perform(NetOperation.InvalidatePools(null)))
        assertEquals(
            NetShellResult.BundlerCacheCleared,
            net.perform(NetOperation.ClearBundlerCache(1)),
        )
    }

    @Test
    fun theDebounceIsARealWait() = runBlocking {
        // The core owns how long; the shell owns only the sleeping. A no-op
        // here would make every keystroke fetch, which is the behaviour the
        // debounce exists to prevent.
        val started = System.nanoTime()
        val result = executor().perform(NetOperation.StartSearchDebounce(40))
        val elapsedMs = (System.nanoTime() - started) / 1_000_000
        assertEquals(NetShellResult.DebounceElapsed, result)
        assertTrue("waited $elapsedMs ms, expected at least 30", elapsedMs >= 30)
    }

    @Test
    fun everyOperationHasANeutralAnswer() {
        // Exhaustiveness is the compiler's (a `when` over a sealed hierarchy
        // with no `else`); this pins the two that could plausibly be given the
        // wrong shape by hand.
        val net = executor()
        assertEquals(
            NetShellResult.StoreLoaded(),
            net.neutralAnswer(NetOperation.ReadStore),
        )
        assertEquals(
            NetShellResult.Probed("u", null, 0),
            net.neutralAnswer(NetOperation.ProbeRpc("u")),
        )
    }
}
